File: 백준/Silver/9012. 괄호/괄호.cc
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <algorithm>
#include <string>
#include <stack>
using namespace std;

int main() {
	int T;
	scanf("%d", &T);

	for (int i = 0; i < T; i++) {
		string input;
		stack<char> check;
		cin >> input;

		string result = "YES";

		for (int j = 0; j < input.size(); j++) {
			if (input[j] == '(') check.push(input[j]);
			else if (check.empty() == false && input[j] == ')') check.pop();
			else {
				result = "NO";
				break;
			}
		}

		if (check.empty() == false) result = "NO";
		cout << result << "\n";
	}
}

File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

T = int(input())

for _ in range(T):
    stack = []
    string = input().strip()
    flag = 0

    for s in string:
        if s == '(':
            stack.append('(')
        else:
            if not stack: 
                flag = 1
                print('NO')
                break
            else:
                stack.pop()
    
    if flag == 0:
        if not stack: print('YES')
        else: print('NO')